addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlCommunication.java

```diff
@@ -0,0 +1,878 @@
+                nhcAddress = nikoHomeControlFindAddr();
+            } else {
+                nhcAddress = nikoHomeControlFindAddr(addr);
+            }
+
+            if (nhcAddress == null) {
+                logger.warn("Niko Home Control: did not find real IP of gateway");
+                return;
+            }
+
+            nhcSocket = new Socket(nhcAddress, nhcPort);
+            nhcOut = new PrintWriter(nhcSocket.getOutputStream(), true);
+            nhcIn = new BufferedReader(new InputStreamReader(nhcSocket.getInputStream()));
+            logger.info("Niko Home Control: connected");
+
+            // initialize all info in local variables
+            initialize();
+
+            // Start Niko Home Control event listener. This listener will act on all messages coming from
+            // IP-interface.
+            (new Thread(nhcEvents)).start();
+
+        } catch (IOException e) {
+            // if the error occurs in the initialization, don't try to restart
+            logger.warn("Niko Home Control: error initializing communication");
+            stopCommunication(false);
+        }
+
+    }
+
+    /**
+     * Runnable that handles inbound communication from Niko Home Control.
+     * <p>
+     * The thread listens to the TCP socket opened at instantiation of the {@link NikoHomeControlCommunication} class
+     * and interprets all inbound json messages. It triggers state updates for active channels linked to the Niko Home
+     * Control actions. It is started after initialization of the communication.
+     *
+     */
+    private Runnable nhcEvents = new Runnable() {
+
+        @Override
+        public void run() {
+            String nhcMessage;
+
+            logger.debug("Niko Home Control: listening for events");
+            listenerStopped = false;
+
+            try {
+                while (!listenerStopped & ((nhcMessage = nhcIn.readLine()) != null)) {
+                    readMessage(nhcMessage);
+                }
+            } catch (IOException e) {
+                if (!listenerStopped) {
+                    // this is not an communication stop triggered from outside this runnable
+                    logger.warn("Niko Home Control: IO error in listener");
+                    // the IO has stopped working, so we need to close cleanly and try to restart
+                    stopCommunication(true);
+                }
+            }
+        }
+
+    };
+
+    /**
+     * Method that interprets all feedback from Niko Home Control and calls appropriate handling methods.
+     *
+     * @param nhcMessage message read from Niko Home Control.
+     */
+    private void readMessage(String nhcMessage) {
+
+        Gson gson = new Gson();
+        NHCCmdAck nhcCmdAck = null;
+        NHCSimpleCmdAck nhcSimpleCmdAck = null;
+
+        logger.debug("Niko Home Control: received json {}", nhcMessage);
+
+        try {
+            // first try with simple reply form, a hashmap in the data field
```

Comment:
I don't like this much myself, but didn't find another easy way of doing this. I cannot derive what I will get from any context, and it is either one of the two forms. I have put this block in a separate method to make it easier to follow.